src/coreclr/vm/jitinterface.cpp

```diff
@@ -2109,15 +2109,19 @@ static GetTypeLayoutResult GetTypeLayoutHelper(
     parNode.hasSignificantPadding = false;
 
     EEClass* pClass = pMT->GetClass();
-    if (pClass->IsNotTightlyPacked() && (!pClass->IsManagedSequential() || pClass->HasExplicitSize() || pClass->IsInlineArray()))
+    if (pClass->IsNotTightlyPacked())
     {
-        // Historically on the JIT side we did not consider types with GC
-        // pointers to have significant padding, even when they have explicit
-        // layout attributes. This retains the more liberal treatment and
-        // lets the JIT still optimize these cases.
-        if (!pMT->ContainsPointers() && pMT != g_TypedReferenceMT)
+        if (pClass->HasExplicitFieldOffsetLayout() || (pClass->IsManagedSequential() && pClass->HasExplicitSize()) ||
+            pClass->IsInlineArray())
         {
-            parNode.hasSignificantPadding = true;
+            // Historically on the JIT side we did not consider types with GC
+            // pointers to have significant padding, even when they have explicit
+            // layout attributes. This retains the more liberal treatment and
+            // lets the JIT still optimize these cases.
```

Comment:
```suggestion
            // Types with GC pointers always use auto layout, irrespective of the explicit layout attribute.
            // Auto layout is internal runtime implementation detail that the user code should not depend on.
            // Thus, padding in auto layout types does not need to be treated as significant.
```
This would be better explanation for why we do it this way.

Reply:
> Types with GC pointers never have significant padding

Is this is a true assumption? Consider:
```cs
ObjX a;
byte b;
((byte*)&a)[10] = 1;
Problem(&a, &b);
Console.WriteLine(b);

[StructLayout(LayoutKind.Explicit, Size = 128)]
struct ObjX
{
    [FieldOffset(0)]
    public object X;
}

[MethodImpl(MethodImplOptions.NoInlining)]
static object Problem(ObjX* pA, byte* pByte)
{
    ObjX a = *pA;
    *pByte = ((byte*)&a)[10];
    return a.X;
}
```
This will print either `1` or `0` depending on whether `Problem`'s optimized or not. It is a fairly long-standing bug.

Reply:
We can decide whether this is a bug or by design. Either way is fine with me.

Reply:
My main worry would be that such struct declarations are common with high performance code trying to avoid false sharing. With that said, it's odd that there should be a difference for structs depending on whether they have GC pointers or not in this regard. I suspect the main reason the check exists is being solved by this PR anyway. Let me collect some diffs with and without the check.

Reply:
FWIW, when I have initially made this change I forgot this check and it resulted in significant regressions; but as mentioned above I suspect those cases are handled by the new less conservative logic.

Reply:
I see no diffs at all over our own frameworks with this special case removed after this change so I will just remove it.